_posts/2023-06-04-not-generating-constructors.md

```diff
@@ -0,0 +1,411 @@
+mov     rdi, rax
+call    std::vector<int, std::allocator<int> >::data()
+mov     rsi, rbx
+mov     rdi, rax
+call    take_pointer(int const*, unsigned long)
+```
+That's a lot more instructions, as there's 2 function calls first. Then store the two values and
+make the call. This is why we had a `const std::string&` overload for those functions. In the case
+the caller already had a decomposed pointer/len, this would result in just something like the last
+three instructions, which is as good as `take_vector()`.
+
+And here's the call to `take_span()`:
+```asm
+lea     rdx, [rbp-80]
+lea     rax, [rbp-48]
+mov     rsi, rdx
+mov     rdi, rax
+call    std::span<int const, 18446744073709551615ul>::span<std::vector<int, std::allocator<int> >&>(std::vector<int, std::allocator<int> >&)
+mov     rdx, QWORD PTR [rbp-48]
+mov     rax, QWORD PTR [rbp-40]
+mov     rdi, rdx
+mov     rsi, rax
+call    take_span(std::span<int const, 18446744073709551615ul>)
+```
+Oof, that's a lot more instructions than `take_vector()`. Calling this function with a vector has
+changed three instructions into ten, and that's _without_ inlining the `std::span` constructor.
+
+I guess it's easy to see why the binary size changes when we starting receiving `std::span` in
+function arguments.
+
+What if we had a pointer and length already, but we wanted to call `take_span()` instead of
+`take_pointer()`, is that any better? Something like this:
+```cpp
+auto p = a.data();
+auto s = a.size();
+take_span({p, s});
+```
+That turns into the following when calling `take_span()`:
+```asm
+mov     rdx, QWORD PTR [rbp-32]
+mov     rcx, QWORD PTR [rbp-24]
+lea     rax, [rbp-64]
+mov     rsi, rcx
+mov     rdi, rax
+call    std::span<int const, 18446744073709551615ul>::span<int*>(int*, unsigned long)
+mov     rdx, QWORD PTR [rbp-64]
+mov     rax, QWORD PTR [rbp-56]
+mov     rdi, rdx
+mov     rsi, rax
+call    take_span(std::span<int const, 18446744073709551615ul>)
+```
+It's one more instruction to call `take_span()` when you have a pointer/length pair than to call
+`take_pointer()` when you have a `std::span`!
+
+Conversions can be costly for codegen in C++, which is why they should pretty much
+[always be explicit](
+https://github.com/chromium/subspace/blob/714e0cec352c77968e3caa3bf78b3b1d83ea984a/PRINCIPLES.md?plain=1#L85-L86),
+even for (if not especially for) these small vocabulary types. I go further to take the point of
+view that conversion constructors are harmful to understanding what code you are generating, even
+if `explicit`, and [static constructor methods are better](
+https://github.com/chromium/subspace/blob/714e0cec352c77968e3caa3bf78b3b1d83ea984a/PRINCIPLES.md?plain=1#L60-L62).
+
+But anyway, what do we do? I want to stop out-of-bounds bugs in C++ code from constantly being
+exploited. I want to stop the people who want to steal all of your private data for the rest of
+your life (which almost certainly you will not know when or if it happens, unless they tell you).

```

Comment:
The "which" and "it" don't clearly refer to any noun in the previous clause (dangling pointers! 🙂); try: "I want to stop the people who want to steal all of your private data. Silent and persistent compromise is not as rare or as difficult as it should be, and that is due in part to memory unsafety..." (or something like that)

Reply:
But anyway, what do we do? I want to stop out-of-bounds bugs in C++ code from constantly being
exploited. I want to stop the people who want to steal all of your private data. Silent and
persistent compromise, which allows theft of your past, present and future data, is not as rare or
as difficult as it should be, and that is due in part to memory unsafety bugs like out of bounds
memory accesses.